Fix parse_qop_value() to not access memory outside of the qop value region

sobomax opened the pull request:
Fix parse_qop_value() to not access memory outside of the qop value region.

Noticed by:	liviuchircu

Use is_ws(), not isspace() while I am here.

Below is the little unit test that I created. I am not sure if there any place to put it in.

```
#include <assert.h>
#include <stddef.h>
#include <stdio.h>
#include <stdlib.h>

#include "parser/parse_authenticate.h"
#include "str.h"

static struct tts {
    const str ts;
    int tres;
} tset[] = {
    {.ts = str_init("auth"), .tres = 0},
    {.ts = str_init("auth-int"), .tres = 0},
    {.ts = str_init("auth,auth-int"), .tres = 0},
    {.ts = str_init("auth, auth-int"), .tres = 0},
    {.ts = str_init("auth       , auth-int"), .tres = 0},
    {.ts = str_init("auth-int,auth"), .tres = 0},
    {.ts = str_init("auth-int  ,                auth"),  .tres = 0},
    {.ts = str_init("gargage"), .tres = -1},
    {.ts = str_init("auth,garbage"), .tres = -1},
    {.ts = str_init("auth , garbage"), .tres = -1},
    {.ts = str_init("auth,,auth-int"), .tres = -1},
    {.ts = str_init("auth-int,"), .tres = -1},
    {.ts = str_init("auth,auth-int,"), .tres = -1},
    {.ts = str_init("auth,auth-int,garbage"), .tres = -1},
    {.ts = str_init("auth-int  ,                auth "),  .tres = -1},
    {.ts = str_init(" auth-int,auth"),  .tres = -1},
    {.tres = 0}
};

int main()
{
    struct tts *tp;
    static struct authenticate_body ab;

    for (tp = tset; tp->ts.len != 0; tp++) {
        int tres = parse_qop_value(tp->ts, &ab);
        printf("parse(\"%s\") = %d\n", tp->ts.s, tres);
        assert(tp->tres == tres);
    }
    return 0;
}

```

```
[sobomax@van01 ~/projects/opensips]$ ./patest
parse("auth") = 0
parse("auth-int") = 0
parse("auth,auth-int") = 0
parse("auth, auth-int") = 0
parse("auth     , auth-int") = 0
parse("auth-int,auth") = 0
parse("auth-int  ,              auth") = 0
parse("gargage") = -1
parse("auth,garbage") = -1
parse("auth , garbage") = -1
parse("auth,,auth-int") = -1
parse("auth-int,") = -1
parse("auth,auth-int,") = -1
parse("auth,auth-int,garbage") = -1
parse("auth-int  ,              auth ") = -1
parse(" auth-int,auth") = -1
```

liviuchircu commented:
hey, @sobomax, I haven't gotten a chance to review this yet, but I'm pushing hard to fit this PR into the current week.  Regarding unit tests, there is already built-in support for both core and module tests, [documented here](https://docs.google.com/document/d/151R7QPADGnzSxtj7FFADGBdbZjsApnkkumiVwl2GmtE/edit?usp=sharing).  I also hope to put up a wiki page with that document's content as well.

Specifically regarding "qop" tests, @rvlad-patrascu already wrote some unit tests, under @parser/test/test_parse_qop.c@ -- please check them out, run them, etc..  If you need any more help regarding the tests, please ask away! :)

sobomax commented:
@liviuchircu thanks for pointers. I've updated the test function to compile, cleaned it up a bit, extended test set and also added checks to make sure memory is not accessed before or after the supplied buffer.

Here is me injecting failure by modifying the first line of the function:

```
ok 76 - parse_qop_value("aut") == -1
ok 77 - oob check: parse_qop_value("")
ok 78 - oob check: parse_qop_value("")
ok 79 - oob check: parse_qop_value("a")
ok 80 - oob check: parse_qop_value("a")
ok 81 - oob check: parse_qop_value("au")
ok 82 - oob check: parse_qop_value("au")
Sep 23 14:01:41 [46609] CRITICAL:core:sig_usr: segfault in attendant (starter) process!
gmake: *** [Makefile.test:17: unit_tests] Segmentation fault (core dumped)
$ gdb91 opensips /var/tmp/cores/opensips.46609.core
Program terminated with signal SIGSEGV, Segmentation fault.
#0  0x000000000052755f in parse_qop_value (val=..., auth=0x7fffffffdf20) at parser/parse_authenticate.c:92
92              if (val.len < 3 || LOWER4B(GET4B(val.s))!=0x61757468) /* "auth" */
(gdb) bt
#0  0x000000000052755f in parse_qop_value (val=..., auth=0x7fffffffdf20) at parser/parse_authenticate.c:92
#1  0x000000000059465c in test_parse_qop_oob (tstr=0x7fffffffdd70, farg=0x7fffffffdf20) at parser/test/test_parse_qop.c:72
#2  0x0000000000595677 in test_oob (sarg=0x7fffffffddb0, tfunc=0x59462a <test_parse_qop_oob>, param=0x7fffffffdf20) at parser/test/test_oob.c:43
#3  0x00000000005945ef in test_parse_qop_val () at parser/test/test_parse_qop.c:64
#4  0x000000000059551a in test_parser () at parser/test/test_parser.c:114
#5  0x00000000005957f3 in run_unit_tests () at test/unit_tests.c:62
#6  0x0000000000467672 in main_loop () at main.c:850
#7  0x000000000046a386 in main (argc=8, argv=0x7fffffffe0f8) at main.c:1491
```

liviuchircu commented on parser/parse_authenticate.c:
```diff
@@ -82,56 +82,63 @@
-	if (*q!=',')
+	if (val.len == 0)
+		return 0;
+	while (val.len > 0 && is_ws(*val.s))
+		STR_ADVANCE(&val);
```
I suggest replacing these two lines with the easy-to-read:

```
trim_leading(&val);
```

liviuchircu commented on parser/parse_authenticate.c:
```diff
@@ -82,56 +82,63 @@
-	while (q<val->s+val->len && isspace((int)*q)) q++;
+postcomma:
+	STR_ADVANCE(&val);
+	while (val.len > 0 && is_ws(*val.s))
+		STR_ADVANCE(&val);
```
```
trim_leading(&val);
```

sobomax replied:
yes! good catch 👍

liviuchircu reviewed (approved):
As promised, managed to look into this.  Everything looks good (especially the mmap tests), except a minor piece of code which could use improvement.  I will fix those myself, as I've already checked out your branch.
